File: AIReaderKindle/src/Services/LibrarySync.cpp
#include "LibrarySync.hpp"

#include "../Domain/Books/BookKey.hpp"
#include "../Domain/Books/RemoteBookName.hpp"
#include "../Support/Files.hpp"
#include "EpubLoader.hpp"
#include "Paths.hpp"
#include "WebDav.hpp"

#include <map>

namespace LibrarySync {

Local gather(Env& env) {
    return {env.library.all(), env.library.remoteNamesMet()};
}

Outcome exchange(const SyncSettings& settings, const Local& local) {
    Outcome outcome;
    std::string folder = settings.booksUrl();
    std::map<std::string, long long> keys;
    std::set<long long> named;
    for (const auto& book : local.books) {
        keys[BookKey::make(book.title, book.author)] = book.id;
        if (!book.remoteName.empty()) named.insert(book.id);
    }

    try {
        std::vector<std::string> remote;
        for (const auto& name : WebDav::list(folder, settings)) {
            if (RemoteBookName::isBook(name)) remote.push_back(name);
        }

        // Fetched before anything is sent, so a book this device already
        // has is recognised and not sent a second time.
        for (const auto& name : remote) {
            if (local.met.count(name)) continue;
            auto contents = WebDav::download(folder + WebDav::escape(name), settings);
            // A file that is not a readable EPUB is met all the same, so it
            // is not fetched again on every sync.
            outcome.met.push_back(name);
            if (!contents) continue;
            std::string file = RemoteBookName::make(Files::stem(name), "", Files::list(Paths::books()));
            std::string path = Files::join(Paths::books(), file);
            if (!Files::write(path, *contents)) throw WebDav::Error("Could not save " + name + " to " + Paths::books() + ".");
            auto metadata = EpubLoader::metadata(path);
            if (!metadata) {
                Files::remove(path);
                continue;
            }
            std::string key = BookKey::make(metadata->title, metadata->author);
            auto same = keys.find(key);
            if (same == keys.end()) {
                keys[key] = 0;
                outcome.received.push_back({name, path});
                continue;
            }
            Files::remove(path);
            if (same->second && !named.count(same->second)) {
                named.insert(same->second);
                outcome.named.push_back({same->second, name});
            }
        }

        std::vector<std::string> taken = remote;
        for (const auto& book : local.books) {
            if (named.count(book.id)) continue;
            auto contents = Files::read(book.path);
            if (!contents) continue;
            std::string name = RemoteBookName::make(book.title, book.author, taken);
            WebDav::upload(folder + WebDav::escape(name), *contents, settings, "application/epub+zip");
            taken.push_back(name);
            outcome.named.push_back({book.id, name});
            outcome.met.push_back(name);
            outcome.sent += 1;
        }
    } catch (const std::exception& failure) {
        outcome.error = failure.what();
    }
    return outcome;
}

void record(Env& env, const Outcome& outcome) {
    for (const auto& name : outcome.met) env.library.meetRemote(name);
    if (!outcome.received.empty()) env.library.refresh(Paths::bookFolders());
    for (const auto& book : env.library.all()) {
        for (const auto& [name, path] : outcome.received) {
            if (book.path == path) env.library.setRemoteName(book.id, name);
        }
    }
    for (const auto& [id, name] : outcome.named) env.library.setRemoteName(id, name);
}

}  // namespace LibrarySync

File: AIReader/AIReader/Services/Sync/LibrarySync.swift
import Foundation
import SQLiteData

/// The books themselves, shared as EPUB files in the `Books` folder beside
/// the sync file. A file there that this device has not met is fetched and
/// shelved; a book here that has no file there yet is sent. The browser
/// extension saves web pages into the same folder.
///
/// Removing a book removes it from this device only: the file stays for the
/// others, and is not fetched again since this device has met it.
///
/// Files are fetched before any are sent, so a book added on two devices
/// separately is recognised by its key and ends up on the server once.
enum LibrarySync {
    struct Outcome: Equatable, Sendable {
        var received = 0
        var sent = 0
    }

    static func run(settings: SyncSettings, database: any DatabaseWriter) async throws -> Outcome {
        guard let folder = settings.booksURL else { return Outcome() }
        var outcome = Outcome()

        let remote = try await list(folder, settings: settings)
        let met = Set(try await database.read { db in try RemoteBook.all.fetchAll(db).map(\.name) })
        for name in remote where !met.contains(name) {
            if try await receive(name, from: folder, settings: settings, database: database) {
                outcome.received += 1
            }
        }

        let unsent = try await database.read { db in try Book.where { $0.remoteName.is(nil) }.fetchAll(db) }
        var taken = Set(remote)
        for book in unsent {
            let name = RemoteBookName.make(title: book.title, author: book.author, avoiding: taken)
            let data = try EPUBPacker.pack(book.directory)
            try await WebDAV.upload(
                data,
                to: folder.appending(path: name),
                type: "application/epub+zip",
                settings: settings
            )
            taken.insert(name)
            try await database.write { db in
                try assign(name, to: book.id, in: db)
                try remember(name, in: db)
            }
            outcome.sent += 1
        }
        return outcome
    }

    /// The EPUB files in the folder; none when there is no folder yet.
    private static func list(_ folder: URL, settings: SyncSettings) async throws -> [String] {
        do {
            return try await WebDAV.list(folder, settings: settings)
                .filter { !$0.isFolder && RemoteBookName.isBook($0.name) }
                .map(\.name)
        } catch WebDAV.DAVError.http(status: 404) {
            return []
        }
    }

    /// Fetches one file and shelves it, unless it is a book this device
    /// already has. True when a book was added.
    private static func receive(
        _ name: String,
        from folder: URL,
        settings: SyncSettings,
        database: any DatabaseWriter
    ) async throws -> Bool {
        guard let data = try await WebDAV.download(folder.appending(path: name), settings: settings) else {
            return false
        }
        // Named as it was on the server, since a book without a title takes
        // its file's name.
        let file = FileManager.default.temporaryDirectory.appending(path: name)
        try data.write(to: file)
        defer { try? FileManager.default.removeItem(at: file) }
        // A file that is not a readable EPUB is met all the same, so it is
        // not fetched again on every sync.
        let unpacked = try? EPUBImporter.unpack(epubAt: file)

        let duplicate = try await database.write { db -> Bool in
            try remember(name, in: db)
            guard let unpacked else { return false }
            let key = BookKey.make(title: unpacked.title, author: unpacked.author)
            if let existing = try Book.all.fetchAll(db).first(where: { $0.key == key }) {
                if existing.remoteName == nil { try assign(name, to: existing.id, in: db) }
                return true
            }
            let remoteName: String? = name
            try Book.insert {
                Book.Draft(
                    title: unpacked.title,
                    author: unpacked.author,
                    language: unpacked.language,
                    folder: unpacked.folder,
                    packagePath: unpacked.packagePath,
                    coverPath: unpacked.coverPath,
                    remoteName: remoteName
                )
            }
            .execute(db)
            return false
        }
        if let unpacked, duplicate {
            try? FileManager.default.removeItem(at: BookStorage.directory(named: unpacked.folder))
        }
        return unpacked != nil && !duplicate
    }

    private static func assign(_ name: String, to bookID: Book.ID, in db: Database) throws {
        let remoteName: String? = name
        try Book.update { $0.remoteName = #bind(remoteName) }.where { $0.id.eq(bookID) }.execute(db)
    }

    private static func remember(_ name: String, in db: Database) throws {
        try RemoteBook
            .insert { RemoteBook(name: name) } onConflict: { $0.name } doUpdate: { row, excluded in
                row.name = excluded.name
            }
            .execute(db)
    }
}
